gpio: add support for get_config and get_direction in the nct38xx driver

barnas-michal opened the pull request:
This PR adds the implementations for the API used to get current configurations of the pins in the NCT38xx IO expanders family.

barnas-michal committed:
gpio: fix the nct38xx driver

This commit fixes the comment and adds the missing assignment of
return value from i2c read byte command in the nct38xx driver.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

cfriedt commented on drivers/gpio/gpio_nct38xx_port.c:
```diff
@@ -6,11 +6,10 @@
 
 #define DT_DRV_COMPAT nuvoton_nct38xx_gpio_port
 
-#include <zephyr/drivers/gpio.h>
```
Whitespace changes (including header re-ordering) are best to be kept in a separate commit.

cfriedt commented on drivers/gpio/gpio_nct38xx_port.c:
```diff
@@ -70,7 +69,7 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
 		}
 
 		new_reg = reg | mask;
-		/* NCT3807 bit3 should be set to 0 */
```
probably should be a separate commit from the gpio dir + config commit

semihalf-barnas-michal replied:
It is separate commit, just the same PR

cfriedt commented on drivers/gpio/gpio_nct38xx_port.c:
```diff
@@ -459,16 +580,16 @@ BUILD_ASSERT(CONFIG_GPIO_NCT38XX_PORT_INIT_PRIORITY > CONFIG_GPIO_NCT38XX_INIT_P
 	static const struct gpio_nct38xx_port_config gpio_nct38xx_port_cfg_##inst = {              \
-		.common = { .port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(inst) &               \
-					     DT_INST_PROP(inst, pin_mask) },                       \
+		.common = {.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(inst) &                \
```
this and the change below are only whitespace. should be in a separate commit

semihalf-barnas-michal replied:
Done

cfriedt reviewed (changes requested):
Looks good - just some nits. Please split this into two commits:
1. commit for adding gpio dir and config
2. whitespace, setting `ret`, comment, and header reordering "clean up"

semihalf-barnas-michal commented on drivers/gpio/gpio_nct38xx_port.c:
```diff
@@ -131,8 +130,8 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
 
 	/* Configure pin as output, if requested 0:input 1:output */
 	if (flags & GPIO_OUTPUT) {
-		nct38xx_reg_read_byte(config->nct38xx_dev, NCT38XX_REG_GPIO_DIR(config->gpio_port),
```
It is in separate commit from the gpio config/dir, it's in the same commit as the change of "should -> must"

barnas-michal committed:
gpio: add support for get_config and get_direction in nct38xx

This commit adds support for get_config and get_direction functions
for the nct38xx IO expander family.
Also applies the clang-format changes.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
gpio: cleanup the nct38xx driver

Reorganize includes and fix the indentation of code.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

cfriedt commented:
@semihalf-barnas-michal - can you please rebase. I'll approve again. 

@henrikbrixandersen - PTAL

sjg20 commented on drivers/gpio/gpio_nct38xx_port.c:
```diff
@@ -146,6 +146,78 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
+	k_sem_take(&data->lock, K_FOREVER);
+
+	if (config->gpio_port == 0) {
+		if (mask & (~config->common.port_pin_mask)) {
```
Please drop the extra brackets as they are confusing

sjg20 commented on drivers/gpio/gpio_nct38xx_port.c:
```diff
@@ -146,6 +146,78 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
+			goto done;
+		}
+
+		if ((mask & config->pinmux_mask) && (mask & (~reg))) {
```
Can simplify to:

mask & config->pinmux_mask & ~reg

How about adding a comment too?

sjg20 commented on drivers/gpio/gpio_nct38xx_port.c:
```diff
@@ -381,6 +453,50 @@ static int gpio_nct38xx_manage_callback(const struct device *dev, struct gpio_ca
+	}
+
+	if (inputs) {
+		*inputs = mask & (~dir_reg);
```
same here - drop brackets
